hal/armv8m/stm32/n6/risaf.c

```diff
@@ -0,0 +1,411 @@
+ * 	 E.g. privCIDMask == 0x02 -> CID 1 can make only privileged accesses, other CIDs can make privileged
+ *   or unprivileged accesses.
+ * `readCIDMask` - Bit mask of which CIDs are permitted to read.
+ * `writeCIDMask` - Bit mask of which CIDs are permitted to write.
+ * `secure` - 0 - Region permits only non-secure accesses, 1 - Region permits only secure accesses.
+ * 	 NOTE: secure == 0 forces all sub-regions to also be non-secure.
+ * `enable` - 0 - Region disabled (default permissions apply instead), 1 - Region enabled
+ */
+int _stm32_risaf_configRegion(
+		unsigned int risaf,
+		u8 region,
+		u32 start,
+		u32 end,
+		u8 privCIDMask,
+		u8 readCIDMask,
+		u8 writeCIDMask,
+		char secure,
+		char enable)
+{
+	const u32 region_offs = (region - 1) * 0x10;
+	u32 tmp, status, lpStatus;
+	if (risaf >= (sizeof(risafs) / sizeof(risafs[0]))) {
+		return -EINVAL;
+	}
+
+	/* Check if the corresponding memory or peripheral is turned on.
+	 * Trying to configure RISAF for a peripheral that is off will result in a crash. */
+	if (risafs[risaf].pctl >= 0) {
+		if (_stm32_rccGetDevClock(risafs[risaf].pctl, &status, &lpStatus) < 0) {
+			return -EINVAL;
+		}
+
+		if (status == 0) {
+			return -ENODEV;
+		}
+	}
+
+	if ((region == 0) || (region > risafs[risaf].n_regions)) {
+		return -EINVAL;
+	}
+
+	if ((start < risafs[risaf].start) || (start > risafs[risaf].end) || (end < risafs[risaf].start) || (end > risafs[risaf].end)) {
+		return -EINVAL;
+	}
+
+	if (((start & risafs[risaf].granularity) != 0) || ((end & risafs[risaf].granularity) != risafs[risaf].granularity)) {
+		return -EINVAL;
+	}
+
+	if (risafs[risaf].isCIDAware == 0) {
+		privCIDMask = (privCIDMask != 0) ? 0xff : 0;
+		readCIDMask = (readCIDMask != 0) ? 0xff : 0;
+		writeCIDMask = (writeCIDMask != 0) ? 0xff : 0;
+	}
+
+	hal_cpuDataMemoryBarrier();
+	tmp = *(risafs[risaf].base + risaf_reg1_cidcfgr + region_offs);
+	tmp &= ~(0xff << 16);
+	tmp |= ((u32)writeCIDMask) << 16;
+	tmp &= ~0xff;
+	tmp |= readCIDMask;
+	*(risafs[risaf].base + risaf_reg1_cidcfgr + region_offs) = tmp;
+	/* Values in registers are not CPU addresses, but offsets within the module's own address space. */
+	start -= risafs[risaf].start;
+	end -= risafs[risaf].start;
+	*(risafs[risaf].base + risaf_reg1_startr + region_offs) = start;
+	*(risafs[risaf].base + risaf_reg1_endr + region_offs) = end;
+	hal_cpuDataMemoryBarrier();
+
+	tmp = *(risafs[risaf].base + risaf_reg1_cfgr + region_offs);
+	if (secure) {
+		tmp |= 1 << 8;
+	}
+	else {
+		tmp &= ~(1 << 8);
+	}
+
+	if (enable) {
+		tmp |= 1;
+	}
+	else {
+		tmp &= ~1;
+	}
+
+	tmp &= ~(0xff << 16);
+	tmp |= ((u32)privCIDMask) << 16;
+	*(risafs[risaf].base + risaf_reg1_cfgr + region_offs) = tmp;
+	hal_cpuDataMemoryBarrier();
+
+	return EOK;
+}
+
+int _stm32_risaf_getFirstDisabledRegion(int risaf)
+{
+	u32 region, region_offs;
+	if (risaf >= (sizeof(risafs) / sizeof(risafs[0]))) {
+		return -EINVAL;
+	}
+
+	for (region = 1; region <= risafs[risaf].n_regions; region++) {
```

Comment:
Nit: better to start with `region = 0` because:
- is more common and thus less confusing,
- right now we need to subtract one on every iteration, otherwise we would only need to add one once ;)

Reply:
That I do get, I mean just the loop. But ok, ack